File: node/src/resources/payouts.ts
import type { PaydClient } from "../client";
import type {
  MpesaPayoutParams,
  MpesaPayoutResponse,
  PanAfricanPayoutParams,
  PanAfricanPayoutResponse,
  MerchantPayoutParams,
  MerchantPayoutResponse,
} from "../types/payouts";
import { validateKenyaPhone, normalizeKenyaPhone, validateInternationalPhone } from "../validators/phone";
import { validateMpesaAmount, validatePositiveAmount } from "../validators/amount";
import { validateRequired, validateEnum } from "../validators/params";

/**
 * Payout (disbursement) operations — send money to recipients.
 */
export class Payouts {
  constructor(private readonly client: PaydClient) {}

  /**
   * Send money to an M-Pesa user in Kenya.
   *
   * @example
   * ```typescript
   * const result = await payd.payouts.mpesa({
   *   phoneNumber: "0700000000",
   *   amount: 500,
   *   narration: "Salary payment",
   *   callbackUrl: "https://example.com/webhook",
   * });
   * console.log(result.transactionReference); // "9BD141203407eW"
   * ```
   */
  async mpesa(params: MpesaPayoutParams): Promise<MpesaPayoutResponse> {
    const callbackUrl = params.callbackUrl || this.client.defaults.callbackUrl;
    const walletType = params.walletType || this.client.defaults.walletType || "local";

    validateRequired({ callbackUrl, narration: params.narration }, [
      "callbackUrl",
      "narration",
    ]);

    const phoneNumber = normalizeKenyaPhone(params.phoneNumber);
    validateKenyaPhone(phoneNumber);
    validateMpesaAmount(params.amount);

    const body: Record<string, unknown> = {
      phone_number: phoneNumber,
      amount: params.amount,
      narration: params.narration,
      callback_url: callbackUrl,
      channel: "MPESA",
      currency: "KES",
    };

    if (walletType !== "local") {
      body.wallet_type = walletType;
    }

    const data = await this.client.request({
      method: "POST",
      path: "/api/v2/withdrawal",
      body,
    });

    return {
      success: data.success as boolean,
      message: data.message as string,
      status: data.status as string,
      transactionReference: resolveTransactionRef(data),
      channel: (data.channel as string) || "",
      amount: (data.amount as number) || 0,
      _raw: data,
    };
  }

  /**
   * Send money to a mobile wallet or bank account across Africa.
   *
   * Requires calling `payd.networks.discover("withdrawal", dialCode)` first
   * to get the networkCode, channelId, providerName, and providerCode.
   *
   * @example
   * ```typescript
   * const networks = await payd.networks.discover("withdrawal", "+234");
   * const opay = networks.findBank("OPay");
   *
   * const result = await payd.payouts.panAfrican({
   *   username: "my_payd_user",
   *   ...opay.toPaymentParams(),
   *   accountName: "bank",
   *   accountHolderName: "John Doe",
   *   accountNumber: "0000000000",
   *   amount: 1800,
   *   phoneNumber: "+2340000000000",
   *   narration: "Payment for services",
   *   currency: "NGN",
   *   callbackUrl: "https://example.com/webhook",
   * });
   * ```
   */
  async panAfrican(params: PanAfricanPayoutParams): Promise<PanAfricanPayoutResponse> {
    const username = params.username || this.client.defaults.username;
    const callbackUrl = params.callbackUrl || this.client.defaults.callbackUrl;
    const walletType = params.walletType || this.client.defaults.walletType || "local";

    validateRequired(
      {
        username,
        callbackUrl,
        narration: params.narration,
        accountName: params.accountName,
        accountHolderName: params.accountHolderName,
        accountNumber: params.accountNumber,
        networkCode: params.networkCode,
        channelId: params.channelId,
        currency: params.currency,
        transactionChannel: params.transactionChannel,
        providerName: params.providerName,
        providerCode: params.providerCode,
      },
      [
        "username",
        "callbackUrl",
        "narration",
        "accountName",
        "accountHolderName",
        "accountNumber",
        "networkCode",
        "channelId",
        "currency",
        "transactionChannel",
        "providerName",
        "providerCode",
      ],
    );

    validateInternationalPhone(params.phoneNumber);
    validatePositiveAmount(params.amount);
    validateEnum(params.accountName, ["bank", "phone"], "accountName");
    validateEnum(params.transactionChannel, ["bank", "phone"], "transactionChannel");

    const body: Record<string, unknown> = {
      username,
      network_code: params.networkCode,
      account_name: params.accountName,
      account_holder_name: params.accountHolderName,
      account_number: params.accountNumber,
      amount: params.amount,
      phone_number: params.phoneNumber,
      channel_id: params.channelId,
      narration: params.narration,
      currency: params.currency,
      callback_url: callbackUrl,
      transaction_channel: params.transactionChannel,
      channel: params.transactionChannel,
      provider_name: params.providerName,
      provider_code: params.providerCode,
    };

    if (walletType !== "local") {
      body.wallet_type = walletType;
    }

    const data = await this.client.request({
      method: "POST",
      path: "/api/v2/payments",
      body,
    });

    return {
      success: data.success as boolean,
      message: data.message as string,
      status: data.status as string,
      transactionReference: resolveTransactionRef(data),
      channel: (data.channel as string) || "",
      amount: (data.amount as number) || 0,
      _raw: data,
    };
  }

  /**
   * Pay to an M-Pesa Paybill or Till number in Kenya.
   *
   * @example
   * ```typescript
   * // Pay a Paybill
   * const result = await payd.payouts.merchant({
   *   username: "my_payd_user",
   *   amount: 500,
   *   phoneNumber: "+254700000000",
   *   narration: "Utility bill payment",
   *   businessAccount: "123456",      // Paybill number
   *   businessNumber: "9876543210",   // Account number
   *   callbackUrl: "https://example.com/webhook",
   * });
   *
   * // Pay a Till
   * const result = await payd.payouts.merchant({
   *   username: "my_payd_user",
   *   amount: 200,
   *   phoneNumber: "+254700000000",
   *   narration: "Store purchase",
   *   businessAccount: "654321",   // Till number
   *   businessNumber: "N/A",       // Not needed for Till
   *   callbackUrl: "https://example.com/webhook",
   * });
   * ```
   */
  async merchant(params: MerchantPayoutParams): Promise<MerchantPayoutResponse> {
    const username = params.username || this.client.defaults.username;
    const callbackUrl = params.callbackUrl || this.client.defaults.callbackUrl;
    const walletType = params.walletType || this.client.defaults.walletType || "local";

    validateRequired(
      {
        username,
        callbackUrl,
        narration: params.narration,
        businessAccount: params.businessAccount,
        businessNumber: params.businessNumber,
      },
      ["username", "callbackUrl", "narration", "businessAccount", "businessNumber"],
    );

    validateInternationalPhone(params.phoneNumber);
    validatePositiveAmount(params.amount);

    const body: Record<string, unknown> = {
      username,
      amount: params.amount,
      currency: "KES",
      phone_number: params.phoneNumber,
      narration: params.narration,
      transaction_channel: "bank",
      channel: "bank",
      business_account: params.businessAccount,
      business_number: params.businessNumber,
      callback_url: callbackUrl,
    };

    if (walletType !== "local") {
      body.wallet_type = walletType;
    }

    const data = await this.client.request({
      method: "POST",
      path: "/api/v3/withdrawal",
      body,
    });

    return {
      success: data.success as boolean,
      message: data.message as string,
      status: data.status as string,
      transactionReference: resolveTransactionRef(data),
      channel: (data.channel as string) || "",
      amount: (data.amount as number) || 0,
      _raw: data,
    };
  }
}

/**
 * Resolve the transaction reference from whichever key the API returns.
 */
function resolveTransactionRef(data: Record<string, unknown>): string {
  return (
    (data.transaction_reference as string) ||
    (data.correlator_id as string) ||
    (data.payd_transaction_ref as string) ||
    ""
  );
}

File: python/src/payd/resources/payouts.py
"""Payout (disbursement) operations."""

from __future__ import annotations

from typing import TYPE_CHECKING

from payd.response import resolve_transaction_reference
from payd.validators import (
    normalize_kenya_phone,
    validate_enum,
    validate_international_phone,
    validate_kenya_phone,
    validate_mpesa_amount,
    validate_positive_amount,
    validate_required,
)

if TYPE_CHECKING:
    from payd.client import PaydClient


class Payouts:
    """Payout (disbursement) operations."""

    def __init__(self, client: PaydClient) -> None:
        self._client = client

    def mpesa(
        self,
        phone_number: str,
        amount: int | float,
        narration: str,
        callback_url: str | None = None,
        wallet_type: str | None = None,
    ) -> dict:
        resolved_callback = callback_url or self._client.defaults.get("callback_url")
        resolved_wallet = wallet_type or self._client.defaults.get("wallet_type") or "local"

        validate_required({"callback_url": resolved_callback, "narration": narration}, ["callback_url", "narration"])
        assert resolved_callback is not None

        normalized_phone = normalize_kenya_phone(phone_number)
        validate_kenya_phone(normalized_phone)
        validate_mpesa_amount(amount)

        body: dict[str, str | int | float] = {
            "phone_number": normalized_phone,
            "amount": amount,
            "narration": narration,
            "callback_url": resolved_callback,
            "channel": "MPESA",
            "currency": "KES",
        }
        if resolved_wallet != "local":
            body["wallet_type"] = resolved_wallet

        data = self._client.request(method="POST", path="/api/v2/withdrawal", body=body)

        return {
            "success": bool(data.get("success")),
            "message": str(data.get("message") or ""),
            "status": str(data.get("status") or ""),
            "transaction_reference": resolve_transaction_reference(data),
            "channel": str(data.get("channel") or ""),
            "amount": float(data.get("amount") or 0),
            "_raw": data,
        }

    def pan_african(
        self,
        account_name: str,
        account_holder_name: str,
        account_number: str,
        network_code: str,
        channel_id: str,
        phone_number: str,
        amount: int | float,
        narration: str,
        currency: str,
        transaction_channel: str,
        provider_name: str,
        provider_code: str,
        username: str | None = None,
        callback_url: str | None = None,
        wallet_type: str | None = None,
    ) -> dict:
        resolved_username = username or self._client.defaults.get("username")
        resolved_callback = callback_url or self._client.defaults.get("callback_url")
        resolved_wallet = wallet_type or self._client.defaults.get("wallet_type") or "local"

        validate_required(
            {
                "username": resolved_username,
                "callback_url": resolved_callback,
                "narration": narration,
                "account_name": account_name,
                "account_holder_name": account_holder_name,
                "account_number": account_number,
                "network_code": network_code,
                "channel_id": channel_id,
                "currency": currency,
                "transaction_channel": transaction_channel,
                "provider_name": provider_name,
                "provider_code": provider_code,
            },
            [
                "username",
                "callback_url",
                "narration",
                "account_name",
                "account_holder_name",
                "account_number",
                "network_code",
                "channel_id",
                "currency",
                "transaction_channel",
                "provider_name",
                "provider_code",
            ],
        )
        assert resolved_username is not None and resolved_callback is not None

        validate_international_phone(phone_number)
        validate_positive_amount(amount)
        validate_enum(account_name, ["bank", "phone"], "account_name")
        validate_enum(transaction_channel, ["bank", "phone"], "transaction_channel")

        body: dict[str, str | int | float] = {
            "username": resolved_username,
            "network_code": network_code,
            "account_name": account_name,
            "account_holder_name": account_holder_name,
            "account_number": account_number,
            "amount": amount,
            "phone_number": phone_number,
            "channel_id": channel_id,
            "narration": narration,
            "currency": currency,
            "callback_url": resolved_callback,
            "transaction_channel": transaction_channel,
            "channel": transaction_channel,
            "provider_name": provider_name,
            "provider_code": provider_code,
        }
        if resolved_wallet != "local":
            body["wallet_type"] = resolved_wallet

        data = self._client.request(method="POST", path="/api/v2/payments", body=body)

        return {
            "success": bool(data.get("success")),
            "message": str(data.get("message") or ""),
            "status": str(data.get("status") or ""),
            "transaction_reference": resolve_transaction_reference(data),
            "channel": str(data.get("channel") or ""),
            "amount": float(data.get("amount") or 0),
            "_raw": data,
        }

    def merchant(
        self,
        amount: int | float,
        phone_number: str,
        narration: str,
        business_account: str,
        business_number: str,
        username: str | None = None,
        callback_url: str | None = None,
        wallet_type: str | None = None,
    ) -> dict:
        resolved_username = username or self._client.defaults.get("username")
        resolved_callback = callback_url or self._client.defaults.get("callback_url")
        resolved_wallet = wallet_type or self._client.defaults.get("wallet_type") or "local"

        validate_required(
            {
                "username": resolved_username,
                "callback_url": resolved_callback,
                "narration": narration,
                "business_account": business_account,
                "business_number": business_number,
            },
            ["username", "callback_url", "narration", "business_account", "business_number"],
        )
        assert resolved_username is not None and resolved_callback is not None

        validate_international_phone(phone_number)
        validate_positive_amount(amount)

        body: dict[str, str | int | float] = {
            "username": resolved_username,
            "amount": amount,
            "currency": "KES",
            "phone_number": phone_number,
            "narration": narration,
            "transaction_channel": "bank",
            "channel": "bank",
            "business_account": business_account,
            "business_number": business_number,
            "callback_url": resolved_callback,
        }
        if resolved_wallet != "local":
            body["wallet_type"] = resolved_wallet

        data = self._client.request(method="POST", path="/api/v3/withdrawal", body=body)

        return {
            "success": bool(data.get("success")),
            "message": str(data.get("message") or ""),
            "status": str(data.get("status") or ""),
            "transaction_reference": resolve_transaction_reference(data),
            "channel": str(data.get("channel") or ""),
            "amount": float(data.get("amount") or 0),
            "_raw": data,
        }
